pkg/proxy/kafka.go

```diff
@@ -0,0 +1,253 @@
+				// Don't report errors while the socket is being closed
+				return
+			default:
+				if err != nil {
+					log.WithFields(log.Fields{
+						"listenPort": redir.conf.listenPort,
+					}).WithError(err).Error("Unable to accept connection")
+					continue
+				}
+			}
+
+			go redir.handleConnection(pair)
+		}
+	}()
+
+	return redir, nil
+}
+
+func (k *kafkaRedirect) canAccess(req *kafka.RequestMessage, srcIdentity uint32) bool {
+	rules := k.rules.GetRelevantRules(k.ingress, uint64(srcIdentity))
+	b, err := json.Marshal(rules.Kafka)
+	out := ""
+	if err != nil {
+		out = err.Error()
+	} else {
+		out = string(b)
+	}
+	log.WithFields(log.Fields{
+		"request": req.String(),
+	}).Debugf("Applying rules %s", out)
+	return req.MatchesRule(rules.Kafka)
+}
+
+func (k *kafkaRedirect) handleRequest(pair *connectionPair, req *kafka.RequestMessage) {
+	log.WithFields(log.Fields{
+		"id":      pair.String(),
+		"request": req.String(),
+	}).Debug("Handling Kafka request")
+
+	addr := pair.rx.conn.RemoteAddr()
+	if addr == nil {
+		log.Warning("RemoteAddr() is nil")
+		return
+	}
+
+	// retrieve identity of source together with original destination IP
+	// and destination port
+	srcIdentity, dstIPPort, err := k.conf.lookupNewDest(addr.String(), k.conf.listenPort)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"id":     pair.String(),
+			"source": addr.String(),
+		}).WithError(err).Error("Unable lookup original destination")
+		return
+	}
+
+	if !k.canAccess(req, srcIdentity) {
+		log.WithFields(log.Fields{
+			"id": pair.String(),
+		}).Debug("Kafka request is denied by policy")
+
+		resp, err := req.CreateResponse(proto.ErrTopicAuthorizationFailed)
+		if err != nil {
+			log.WithFields(log.Fields{
+				"id": pair.String(),
+			}).WithError(err).Error("Unable create response message")
+			return
+		}
+
+		pair.rx.Enqueue(resp.GetRaw())
+		return
+	}
+
+	marker := 0
+	if !k.conf.noMarker {
+		marker = GetMagicMark(k.ingress) | int(srcIdentity)
+	}
+
+	log.WithFields(log.Fields{
+		"id":          pair.String(),
+		"marker":      marker,
+		"destination": dstIPPort,
+	}).Debug("Dialing original destination")
+
+	txConn, err := ciliumDialer(marker, addr.Network(), dstIPPort)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"id":          pair.String(),
+			"origNetwork": addr.Network(),
+			"origDest":    dstIPPort,
+		}).WithError(err).Error("Unable to dial original destination")
+		return
+	}
+
+	pair.tx.conn = txConn
+	pair.tx.startWriter()
+	pair.tx.startReadAndPipe(pair.rx)
+
+	log.WithFields(log.Fields{
+		"id": pair.String(),
+	}).Debug("Forwarding Kafka request")
+
+	// Write the entire raw request onto the outgoing connection
+	pair.tx.Enqueue(req.GetRaw())
+}
+
+func (k *kafkaRedirect) handleConnection(pair *connectionPair) {
+	log.WithFields(log.Fields{
+		"from": pair.rx,
+		"to":   pair.tx,
+	}).Debug("Handling new Kafka connection")
+
+	for {
+		req, err := kafka.ReadRequest(pair.rx.conn)
+		if err != nil {
+			if err == io.EOF {
```

Comment:
This is the only condition that makes this infinite loop to break? Are we sure this doesn't loop for ever in an event of a error that is not `io.EOF`?

Reply:
Other suggestions? The parser always consumes n bytes so the reader progresses until EOF. Added `io.UnexpectedEOF`. The parser can't stop on generic parser errors.